# WASP v4.19 (20260130)

Processing and Analysis of Single-cell RNA-Sequencing and ATAC-Sequencing Datasets

## Description

Incorporates Seurat and Signac with various R packages to perform processing and analysis of single-cell RNA-Sequencing (scRNA-Seq), single-cell ATAC-Sequencing (scATAC-Seq), or combined 10X Multiome datasets.    The functions included in this package provide a comprehensive workflow for integrating commonly used Seurat and Signac functions with advanced statistical methods and visualization.    Most analyses can be run in parallel using intuitive functions to expedite time-consuming steps such as dataset integration and differential expression analysis.    The package is compatible with Windows, Linux, or WSL2. However, analyses conducted in Windows default to sequential processing due to inherent stability issues of parallel processing in Windows.

* Important: Functions run in Linux or WSL2 require installation of additional dependencies and setup of a conda environment. PhenoCycler analysis functions have been removed from this package and are now available as a standalone package at cschasestevens/STIngPC.

## Getting Started

### Dependencies
* Windows 10-11, WSL Ubuntu 22.04 or higher, Linux Ubuntu 22.04 or higher, or macOS 12.7.1 or higher
* R version 4.5.0 or higher (https://cran.r-project.org/)
* (Optional) RStudio version 2023.06.2 or higher (https://posit.co/download/rstudio-desktop/)
* R-packages (downloaded from CRAN unless otherwise specified):
    * Suggests: 
        * knitr,
        * rmarkdown,
        * reticulate,
        * fgsea (Bioconductor),
        * scDblFinder (Bioconductor),
        * SingleCellExperiment (Bioconductor),
        * SummarizedExperiment (Bioconductor),
        * SoupX,
        * parallel,
        * decontX (Bioconductor),
        * plotly,
        * htmlwidgets,
        * patchwork,
        * Azimuth,
        * BSgenome.Hsapiens.UCSC.hg38 (Bioconductor),
        * chromVAR,
        * JASPAR2020 (Bioconductor),
        * TFBSTools (Bioconductor),
        * MAST (Bioconductor),
        * org.Hs.eg.db (Bioconductor),
        * CellChat,
        * biomaRt (Bioconductor),
        * topGO (Bioconductor)
    * Imports: 
        * ggplot2,
        * viridis,
        * ggsci,
        * rtracklayer,
        * GenomeInfoDb (Bioconductor),
        * Rsamtools (Bioconductor),
        * Seurat,
        * SeuratObject,
        * BiocGenerics (Bioconductor),
        * Signac,
        * BSgenome (Bioconductor),
        * harmony,
        * future,
        * ggpubr,
        * ggrepel,
        * dplyr,
        * reshape2,
        * magrittr,
        * circlize,
        * ComplexHeatmap (Bioconductor),
        * grid,
        * gtools,
        * EnhancedVolcano (Bioconductor),
        * shadowtext,
        * stringr,
        * motifmatchr (Bioconductor)

### Installation
* Run the following in a new R session on the command line or within R-Studio:

```
devtools::install_github(
  "cschasestevens/WASP", 
  ref = "main", 
  build_vignettes = TRUE
)
```

## Help
* Browse vignettes by running the following:

```
browseVignettes("WASP")
```

* Access function documentation by running the following:

```
# Type function name after package name
?WASP::sc_umap()
```

## Authors

* Nathanial Chase Stevens, PhD, University of North Carolina at Chapel Hill
* Email: Nathanial_Stevens@med.unc.edu
* Alternate email: cschasestevens@gmail.com
* LinkedIn: https://www.linkedin.com/in/nathanial-chase-stevens-phd-08775180/

## Version History
* 4.19
    * Bioconductor dependencies are now automatically installed along with package.
    * Added option to plot DGEA/DA results with sc_heatmap function.
* 4.18
    * Fixed sc_predict integration with sc_recluster where seurat_clusters is selected instead of the actual reclustering results.
    * Fixed multiome reclustering within sc_recluster to include the clustering resolution parameter.
* 4.17
    * Removed Seurat VlnPlot function from package functions to resolve PackageCheck() error derived from Seurat.
    * Fixed issue in sc_heatmap to include annotated transcription factor names in the final plot.
    * Updated sc_recluster documentation.
    * Fixed compatibility issue in scparam related to different column names used in formatting the reference .gtf file.
* 4.16
    * Updated function documentation.
    * Updated tutorial to v4.16.
    * Integrated fgsea-based enrichment analyses into sc_enrichment function.
    * Updated parameters for sc_enrichment function.
    * Added sc_enrich_plot function for plotting results generated by sc_enrichment.
    * Removed sc_plot_enrichment, which is superceded by sc_enrich_plot.
* 4.15
    * Updated package tutorial to v4.15.
    * Updated sc_coverage_plot parameters and simplified input.
    * Removed sc_coverage_plot2 function.
    * Updated sc_dotplot documentation.
* 4.14
    * Updated package tutorial to v4.14.
    * Updated imports for sc_process functions.
    * Simplified sc_diff input parameters, updated function documentation, and fixed error where missing transcription factors/genes were not included in sc_diff output.
* 4.13
    * Integrated cell type predictions into sc_recluster.
    * Simplified output of sc_recluster and updated input parameters.
    * Fixed error in sc_predict where an existing "predicted.id" column prevented the addition of new cell type predictions to a Seurat object.
    * Revised sc_cc_run to be compatible with Windows.
    * Updated parameters for sc_cc_chrd.
* 4.12
    * Added option to include labels in sc_barplot.
    * Updated marker gene table and fixed error. in selecting top DEGs for sc_heatmap.
    * Added motifmatchr to imports.
    * Removed sc_heatmap_deg and sc_top10_marker_heatmap_rc, which have been superceded by sc_heatmap.
    * Removed sc_seqanno_plot.
* 4.11
    * Fixed namespace
    * Removed redundant packages
* 4.10
    * Updated tutorial to v4.10
    * Updated sc_multiome_params to enable directory selection.
    * Added progress messages to major processing functions.
    * Updated sc_multiome_process and sc_multiome_process_batch to be compatible with MACS3.
    * Updated sc_integrate_data with stability improvements for multiome and scRNA-Seq data integration
    * Removed support for Seurat versions older than 5* with sc_integrate_data.
    * Integrated sc_umap with sc_umap_panel.
    * Removed sc_umap_panel_gene, sc_umap_panel_gene_list, and sc_umap_standard functions, which have been superceded by sc_umap.
    * Updated sc_violin to accept gene names or metadata column names.
    * Fixed bug in sc_predict where wrong cell type assignments are used to label clusters.
* 4.03
    * Updated sc_heatmap() to accept a custom gene list in addition to the default option of calculating the top marker genes per cluster.
    * Updated sc_recluster() function with default parameters.
    * Fixed error in sc_recluster() function where specific reduction names were required to perform reclustering.
    * Revised sc_diff() function for performance improvements.
    * Added function within sc_violin() to stratify plot by a metadata variable.

* 4.02
    * Added generic heatmap function
* 4.01
    * Simplified standard UMAP function
    * Removed code for automatically saving plots generated by sc_recluster()
    * Added generic violin plot function
* 4.00
    * Removed functions for PhenoCycler analysis
* 3.08
    * Integrated spatial proteomics visualizations with existing plot functions.
* 3.07
    * Updated Phenocycler quality control functions.
* 3.06
    * Added parameters for accepting genomic coordinates as inputs for sc_coverage_plot function.
    * Fixed errors in sc_coverage_plot function.
* 3.05
    * Revised phenocycler QC function.
    * Bug fixes for conversion between phenocycler file types.
    * Updated phenocycler normalization function.
    * Updated package documentation
    * Moved scATAC-Seq libraries from "Imports" to "Suggests"
* 3.04
    * Fixed errors when using sc_diff() for differential activity analysis.
    * Updated plot themes.
* 3.03
    * Added function for normalizing PC image signal intensities
    * Added function for generating a QC summary of a PC dataset
    * Added function for converting Seurat objects to AnnData files
* 3.02
    * Added functions for file conversion from Seurat to .h5ad.
    * Added function for visualizing phenocycler cell maps.
    * Integrated phenocycler data format into existing functions for scRNA-Seq and multiome data analysis.
* 3.00
    * Added functions for QC and analysis of Phenocycler segmentation results.
* 2.4
    * Bug fixes for subsetting and reclustering multiome data.
    * Simplified parameters for UMAP and visualization.
    * Added function for Azimuth based cell type predictions.
    * Updated custom cell type prediction function.
    * Updated dot plot function to accept both gene expression and ATAC data.
* 2.3
    * Simplified plotting functions with default parameters.
    * Added batch plotting functions for analyses spanning multiple Seurat objects.
* 2.2
    * Integration of Cell Chat analysis and chord diagram plotting functions.
    * Revised scATAC-Seq coverage plot functions for greater plotting flexibility.
* 2.1
    * Added support for reclustering analysis of scRNA-Seq data
    * Added flexibility for heat map and UMAP functions
    * Addresses errors in package documentation
* 2.0
    * Added functions for processing 10X multiome data sets
    * Updated v1.2 functions for compatibility with Seurat v5.0 and higher
    * Provided additional flexibility for plot parameters
    * Added common multiome plot types, including coverage plots and gene track plots
* 1.2
    * Added functions for enrichment analysis (both Gene Ontology and custom gene sets)
* 1.0
    * Initial Release

## License

This project is licensed under the MIT License - see the LICENSE.md file for details

## Acknowledgments

* Seurat package: Hao, Y., Stuart, T., Kowalski, M.H. et al. Dictionary learning for integrative, multimodal and scalable single-cell analysis. Nat Biotechnol 42, 293–304 (2024). https://doi.org/10.1038/s41587-023-01767-y
* ComplexHeatmap package: Gu Z, Eils R, Schlesner M (2016). “Complex heatmaps reveal patterns and correlations in multidimensional genomic data.” Bioinformatics. <doi:10.1093/bioinformatics/btw313>
* Circlize package: Gu, Z. circlize implements and enhances circular visualization in R. Bioinformatics 2014.
