---
title: "WASP_tutorial_v4-01"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{WASP_tutorial_v4-00}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Re-load previous session objects for quick analysis (if already processed and annotated)
```{r inst, echo=T, results=T, message=F, warning=F,eval=FALSE}
#---- Load data and session objects (if already processed and annotated) ----
# Set working directory
# Load libraries
# Load RDS
# Load gene lists
```

## Data processing
```{r data-proc, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}

```

## CellType Prediction
```{r cell-pred, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}

```

## Analyze a Reclustered Subset of scRNA-Seq or Multiome Data
### Reclustering w/DGEA and DA analysis
```{r reclus, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}

```

### Visualization of Reclustering: CellType and Expression UMAPs
```{r reclus2, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}

```

### Cell Type Predictions for Reclustered Data
```{r pred2, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}

```

## Statistical Analysis
### Differential Gene Expression/Accessibility Analysis (reclustered cell types)
```{r dgea, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}

```

## Coverage Plots
```{r covg, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}

```

## Dot Plots
```{r dotp, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}

```

## Gene Ontology/Custom Gene Set Enrichment Analysis **(Development)**
```{r go-enrich, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}

```

## CellChat Analysis
```{r cchat, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}

```
