---
title: "WASP_tutorial_v4-10"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{WASP_tutorial_v4-10}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Setup
### - Ensure that a 10X CellRanger output folder containing both RNA and ATAC data is available in the working directory prior to running package functions.
### - Also requires a valid annotation and reference genome file for mapping reads/fragments
###   - Recommended to use files from GENCODE (works natively with sc_multiome_params())
### - Example data are available from the 10X website.
```{r inst, echo=T, results=T, message=F, warning=F,eval=FALSE}
#---- Load data and session objects (if already processed and annotated) ----
# Set working directory with setwd()
## Change to relative working directory path on local machine
getwd()
setwd("/home/ncsteven/analyze/Hiro.CF.multiome/WASP_test")
# Load libraries
library(WASP)
library(Seurat)
library(JASPAR2020)
library(TFBSTools)
library(chromVAR)
library(CellChat)
library(BSgenome.Hsapiens.UCSC.hg38)
# Set parameters with sc_multiome_params (or sc_proc_params for scRNA-Seq only)
## - data frame containing sampleID (matches CellRanger folder name)
##   and group info
## - path to genome annotation file
## - path to genome reference file
scparam <- sc_multiome_params(
  study_md = data.frame(
    "Code" = c("DD023T", "KK010T"),
    "Group" = factor(c("NonCF", "CF"), levels = c("NonCF", "CF"))
  ),
  gtf_path = "gencode.v45.primary_assembly.annotation.gtf",
  fa_path = "GRCh38.primary_assembly.genome.fa"
)
```

## Data processing and Integration
```{r data-proc, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}
#---- Processing of 10X multiome files ----
# Process example file using sc_multiome_process_batch()
## Run ?sc_multiome_process_batch() for a full list of processing params
## Need to install MACS3 and hdf5r in a conda environment outside of R
## MACS3 path MUST be specified as CallPeaks will only search for older
## MACS2 installation.
scdata <- sc_process_multiome_batch(scparam, parl = TRUE, core_num = 2)
gc(reset = TRUE)
ls()
#---- Data Integration ----
scint <- sc_integrate_data(
  l_so = scdata,
  l_par = scparam,
  proc_mode = "multiome",
  parl = TRUE,
  core_num = 2
)
#---- Integration QC ----
## Visualize Clusters
sc_umap_panel(
  so = scint,
  md_list = c("Group", "Code", "seurat_clusters"),
  slot1 = "wnn.umap",
  pos_leg = c(0.25, 0.85),
  show_lab = FALSE
)
sc_violin(
  so = scint,
  ct_col = "seurat_clusters",
  gene_col = c(
    "nFeature_RNA",
    "nCount_RNA",
    "percent.mt",
    "nCount_ATAC",
    "TSS.enrichment",
    "nucleosome_signal"
  )
)
# Save integrated data
saveRDS(scint, "processing/int_final.rds")
```

## CellType Prediction
```{r cell-pred, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}
#---- Azimuth Cell Type Prediction ----
scpred <- sc_predict(
  so = scint,
  md_list = c("Code"),
  f_size = 30000
)
# Check assignments
head(scpred[["data"]]@meta.data)
sc_umap_panel(
  so = scpred[["data"]],
  md_list = c("seurat_clusters", "predicted.id"),
  slot1 = "wnn.umap",
  show_lab = TRUE
)
# Check proportions
sc_barplot(
  dat = scpred[["predictions"]],
  ct_col = "predicted.ann_finest_level",
  xvar = "seurat_clusters"
)
# Check marker genes per cluster
sc_heatmap(
  so = scpred[["data"]],
  cl_var = "seurat_clusters"
)
# Save data
saveRDS(scpred, "processing/anno_assigned.rds")
```

## CellType Annotation
```{r cell-anno, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}
#---- Manually assign cell types based on prediction ----
# Manually assign existing CellType column using consensus IDs
cts <- dplyr::select(
  dplyr::left_join(
    scpred[["data"]]@meta.data["seurat_clusters"],
    data.frame(
      "seurat_clusters" = paste(
        gtools::mixedsort(
          levels(scpred[["data"]]@meta.data[["seurat_clusters"]])
        )
      ),
      "CellType" = factor(
        c("0.AT2", "1.Secretory", "2.Endothelial",
          "3.Ciliated", "4.Monocytes", "5.T Cells",
          "6.Alveolar macrophages", "7.AT2",
          "8.DC2", "9.AT1", "10.AT2",
          "11.Basal", "12.AT2", "13.Pericytes",
          "14.AT2", "15.AT2", "16.AT2",
          "17.B cells", "18.Monocytes", "19.AT2",
          "20.AT1", "21.AT2",
          "22.AT2", "23.Endothelial", "24.AT1"
        ),
        levels = c("0.AT2", "1.Secretory", "2.Endothelial",
          "3.Ciliated", "4.Monocytes", "5.T Cells",
          "6.Alveolar macrophages", "7.AT2",
          "8.DC2", "9.AT1", "10.AT2",
          "11.Basal", "12.AT2", "13.Pericytes",
          "14.AT2", "15.AT2", "16.AT2",
          "17.B cells", "18.Monocytes", "19.AT2",
          "20.AT1", "21.AT2",
          "22.AT2", "23.Endothelial", "24.AT1"
        )
      )
    ),
    by = "seurat_clusters"
  ),
  .data[["CellType"]]
)
scanno <- SeuratObject::AddMetaData(
  scpred[["data"]],
  cts[["CellType"]],
  col.name = "CellType"
)
# UMAP with final cluster assignments
sc_umap(
  so = scanno,
  md_var = "CellType",
  slot1 = "wnn.umap"
)
## Save
saveRDS(scanno, "processing/anno_final.rds")
```

## Add Activity Scores to A Multiome Dataset
```{r add-act, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}
## Add motifs
d_chrmasy <- sc_atac_motifs(
  so = scanno,
  dref = scparam[["ref.gtf"]]
)
## Add chromassay to Seurat
scanno[["TF"]] <- d_chrmasy
## Calculate Chromvar activity scores
Seurat::DefaultAssay(scanno) <- "TF"
scanno <- Signac::RegionStats(
  scanno,
  genome = BSgenome.Hsapiens.UCSC.hg38
)
scanno <- Signac::RunChromVAR(
  object = scanno,
  genome = BSgenome.Hsapiens.UCSC.hg38
)
Seurat::DefaultAssay(scanno) <- "chromvar"
saveRDS(scanno, "analyze/scdata.rds")
```

## Analyze a Reclustered Subset of scRNA-Seq or Multiome Data
### Reclustering w/DGEA and DA analysis, cell type predictions
```{r reclus, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}
# Load Seurat object and subset
scre <- readRDS("analyze/scdata.rds")
levels(scre@meta.data[["CellType"]])
scre2 <- scre[
  ,
  grepl(
    "Ciliat|Secret|Basal|AT1|AT2",
    scre@meta.data[["CellType"]]
  )
]
# Recluster subset
scre2 <- sc_recluster(
  so = scre2,
  rc_type = "Mult",
  slot1 = "SCT",
  md_list = "Code",
  cor_col = "Code"
)
names(scre2)
levels(scre2[["predictions.predicted.id"]])
# Assign final cell types
cts <- dplyr::select(
  dplyr::left_join(
    scre2[["data"]]@meta.data["seurat_clusters"],
    data.frame(
      "seurat_clusters" = paste(
        gtools::mixedsort(
          levels(scre2[["data"]]@meta.data[["seurat_clusters"]])
        )
      ),
      "CellType" = factor(
        c("0.AT2", "1.Secretory", "2.Ciliated",
          "3.AT2", "4.AT1", "5.AT2",
          "6.Basal", "7.AT2",
          "8.AT2", "9.AT2", "10.AT2",
          "11.AT2", "12.AT1", "13.AT2",
          "14.AT2", "15.AT1"
        ),
        levels = c("0.AT2", "1.Secretory", "2.Ciliated",
          "3.AT2", "4.AT1", "5.AT2",
          "6.Basal", "7.AT2",
          "8.AT2", "9.AT2", "10.AT2",
          "11.AT2", "12.AT1", "13.AT2",
          "14.AT2", "15.AT1"
        )
      )
    ),
    by = "seurat_clusters"
  ),
  .data[["CellType"]]
)
scre2 <- SeuratObject::AddMetaData(
  scre2[["data"]],
  cts[["CellType"]],
  col.name = "CellType"
)
# UMAP with final cluster assignments
sc_umap(
  so = scre2,
  md_var = "CellType",
  slot1 = "re_wnn_umap"
)
## Save
saveRDS(scre2, "analyze/scdata_reclus.rds")
```

### Visualization of Reclustering: CellType and Expression UMAPs
```{r reclus2, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}
# Visualize expression/accessibility for a specific gene
sc_panel_gex(
  so = scre2,
  slot1 = "re_wnn_umap",
  var_md = "SCGB3A2"
)
```

## Statistical Analysis
### Differential Gene Expression/Accessibility Analysis
```{r dgea, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}
# Load data
d1 <- readRDS("analyze/scdata_reclus.rds")
# DGEA
## Perform test
scdgea <- sc_diff(
  so = d1,
  mast_comp = "GroupNonCF",
  mast_name = "NonCF vs. CF",
  form1 = c("Group", "nFeature_SCT"),
  cores = 4
)
## Save results
lapply(
  seq.int(1, length(scdgea), 1),
  function(x) {
    write.table(
      scdgea[[x]],
      file = paste(
        "analyze/dgea_",
        names(scdgea)[[x]],
        ".txt",
        sep = ""
      ),
      sep = "\t",
      row.names = FALSE
    )
    print(paste(names(scdgea)[[x]], "saved!"))
  }
)
# DA
## Perform test
scdact <- sc_diff(
  so = d1,
  mast_comp = "GroupNonCF",
  mast_name = "NonCF vs. CF",
  form1 = c("Group", "nFeature_SCT"),
  cores = 4,
  asy = "chromvar",
  slot1 = "data"
)
head(scdact[["results"]])
## Save results
lapply(
  seq.int(1, length(scdact), 1),
  function(x) {
    write.table(
      scdact[[x]],
      file = paste(
        "analyze/da_",
        names(scdact)[[x]],
        ".txt",
        sep = ""
      ),
      sep = "\t",
      row.names = FALSE
    )
    print(paste(names(scdact)[[x]], "saved!"))
  }
)
```

## Volcano Plots
```{r vol, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}
# Plot DGEA and DA results
## DGEA
sc_volcano(
  l_deg = scdgea[["results"]],
  ct = "1.Secretory",
  comp_name = "NonCF vs. CF",
  x_title = paste(
    "Higher Activity in CF",
    "<----->",
    "Higher Activity in nonCF",
    sep = " "
  )
)
## DA
sc_volcano(
  l_deg = scdact[["results"]],
  ct = "1.Secretory",
  comp_name = "NonCF vs. CF",
  gene_name = "TF",
  x_title = paste(
    "Higher Activity in CF",
    "<----->",
    "Higher Activity in nonCF",
    sep = " "
  )
)
```

## Coverage Plots
```{r covg, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}
# Generate coverage plot for genomic region closest to
# selected gene transcription site
sc_coverage_plot(
  so = d1,
  dref = scparam[["ref.gtf"]],
  g_name = "SCGB1A1"
)
```

## Dot Plots
```{r dotp, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}
# Generate dotplot from dgea/da list or from custom gene list
sc_dotplot(
  so = d1,
  l_gene = c("SFTPB", "RNASE1", "FOXA2", "TMEM45A", "MUC5B"),
  ct_var = "CellType"
)
```

## Gene Ontology/Custom Gene Set Enrichment Analysis **(Development)**
```{r go-enrich, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}

```

## CellChat Analysis
```{r cchat, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}
# Run CellChat
scchat <- sc_cc_run(
  so = scre2,
  g_name = "Group",
  s_name = "Code",
  parl = FALSE
)
# Format as data frame and merge
scchat2 <- list(
  "LR" = dplyr::bind_rows(
    lapply(
      seq.int(1, length(scchat), 1),
      function(x) {
        d1 <- CellChat::subsetCommunication(
          scchat[[x]],
          # Switch to "netP" for pathways
          slot.name = "net"
        )
        d1[["Group"]] <- rep(names(scchat)[[x]], nrow(d1))
        return(d1) # nolint
      }
    )
  ),
  "PW" = dplyr::bind_rows(
    lapply(
      seq.int(1, length(scchat), 1),
      function(x) {
        d1 <- CellChat::subsetCommunication(
          scchat[[x]],
          # Switch to "netP" for pathways
          slot.name = "netP"
        )
        d1[["Group"]] <- rep(names(scchat)[[x]], nrow(d1))
        return(d1) # nolint
      }
    )
  )
)
saveRDS(scchat2, "analyze/scdata_cellchat.rds")
write.table(
  scchat2[[1]],
  "analyze/scdata_cellchat_LR.txt",
  sep = "\t",
  row.names = FALSE
)
write.table(
  scchat2[[2]],
  "analyze/scdata_cellchat_PW.txt",
  sep = "\t",
  row.names = FALSE
)
# Chord diagrams
head(scchat2[[1]])
levels(scchat2[[1]][["source"]])
sc_cc_chrd(
  ccdf = scchat2[[1]],
  title1 = "plot_cellchat_LR",
  title2 = "L-R Interactions: Subset",
  cc_type = "comp",
  col_g = "Group"
)
```
